src/library/equations_compiler/util.cpp

```diff
@@ -122,11 +122,11 @@ unpack_eqns::unpack_eqns(type_context & ctx, expr const & e):
             }
         } else {
             /* noequation, guess arity using type of function */
-            expr type = ctx.infer(m_fns[fidx]);
+            expr type = ctx.relaxed_whnf(ctx.infer(m_fns[fidx]));
             unsigned arity = 0;
             while (is_pi(type)) {
                 arity++;
-                type = binding_body(type);
+                type = ctx.relaxed_whnf(binding_body(type));
             }
```

Comment:
Type inference methods assume we don't have dangling de-Bruijn variables.
`whnf` depends on type inference (example K-rule in the kernel).

All `whnf` (and type inference) methods assume we don't have dangling de-Bruijn variables.

The following artificial example produces an error:
```lean
def f := ∀ (a : nat) (h : a = a), @eq.rec nat a (λ x, Prop) (empty → false) a h

lemma ex : f.
```